ref/go/src/bech32/bech32.go

```diff
@@ -0,0 +1,193 @@
+// Copyright (c) 2017 Takatoshi Nakagawa
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package bech32
+
+import (
+	"fmt"
+	"strings"
+)
+
+var CHARSET string = "qpzry9x8gf2tvdw0s3jn54khce6mua7l"
+
+var GENERATOR []int = []int{0x3b6a57b2, 0x26508e6d, 0x1ea119fa, 0x3d4233dd, 0x2a1462b3}
+
+func polymod(values []int) int {
+	chk := 1
+	for _, v := range values {
+		top := chk >> 25
+		chk = (chk&0x1ffffff)<<5 ^ v
+		for i := 0; i < 5; i++ {
+			if (top>>uint(i))&1 == 1 {
+				chk ^= GENERATOR[i]
+			}
+		}
+	}
+	return chk
+}
+
+func hrpExpand(hrp string) []int {
+	ret := []int{}
+	for _, c := range hrp {
+		ret = append(ret, int(string(c)[0]>>5))
+	}
+	ret = append(ret, 0)
+	for _, c := range hrp {
+		ret = append(ret, int(string(c)[0]&31))
+	}
+	return ret
+}
+
+func verifyChecksum(hrp string, data []int) bool {
+	return polymod(append(hrpExpand(hrp), data...)) == 1
+}
+
+func createChecksum(hrp string, data []int) []int {
+	values := append(append(hrpExpand(hrp), data...), []int{0, 0, 0, 0, 0, 0}...)
+	mod := polymod(values) ^ 1
+	ret := []int{}
+	for p := 0; p < 6; p++ {
+		ret = append(ret, (mod>>uint(5*(5-p)))&31)
```

Comment:
if you allocate `ret := make([]int, 6)`, you can just assign it directly: `ret[p] = ...`.